File: content/manual/versions/v7.2/source/includes/qe-tutorials/python/queryable_encryption_tutorial.py
from pymongo import MongoClient
import queryable_encryption_helpers as helpers
import os
from dotenv import load_dotenv

load_dotenv()

# start-setup-application-variables
# KMS provider name should be one of the following: "aws", "gcp", "azure", "kmip" or "local"
kms_provider_name = "<KMS provider name>"

uri = os.environ['MONGODB_URI']  # Your connection URI

key_vault_database_name = "encryption"
key_vault_collection_name = "__keyVault"
key_vault_namespace = f"{key_vault_database_name}.{key_vault_collection_name}"
encrypted_database_name = "medicalRecords"
encrypted_collection_name = "patients"
# end-setup-application-variables

kms_provider_credentials = helpers.get_kms_provider_credentials(
    kms_provider_name)
customer_master_key_credentials = helpers.get_customer_master_key_credentials(
    kms_provider_name)

auto_encryption_options = helpers.get_auto_encryption_options(
    kms_provider_name,
    key_vault_namespace,
    kms_provider_credentials,
)

# start-create-client
encrypted_client = MongoClient(
    uri, auto_encryption_opts=auto_encryption_options)
# end-create-client

encrypted_client[key_vault_database_name][key_vault_collection_name].drop()
encrypted_client[encrypted_database_name][encrypted_collection_name].drop()

# start-encrypted-fields-map
encrypted_fields_map = {
    "fields": [
        {
            "path": "patientRecord.ssn",
            "bsonType": "string",
            "queries": [{"queryType": "equality"}]
        },
        {
            "path": "patientRecord.billing",
            "bsonType": "object",
        }
    ]
}
# end-encrypted-fields-map

client_encryption = helpers.get_client_encryption(
    encrypted_client,
    kms_provider_name,
    kms_provider_credentials,
    key_vault_namespace
)

try:
    # start-create-encrypted-collection
    client_encryption.create_encrypted_collection(
        encrypted_client[encrypted_database_name],
        encrypted_collection_name,
        encrypted_fields_map,
        kms_provider_name,
        customer_master_key_credentials,
    )
    # end-create-encrypted-collection
except Exception as e:
    raise Exception("Unable to create encrypted collection due to the following error: ", e)

# start-insert-document
patient_document = {
    "patientName": "Jon Doe",
    "patientId": 12345678,
    "patientRecord": {
        "ssn": "987-65-4320",
        "billing": {
            "type": "Visa",
            "number": "4111111111111111",
        },
    },
}

encrypted_collection = encrypted_client[encrypted_database_name][encrypted_collection_name]

result = encrypted_collection.insert_one(patient_document)
# end-insert-document
if result.acknowledged:
    print("Successfully inserted the patient document.")


# start-find-document
find_result = encrypted_collection.find_one({
    "patientRecord.ssn": "987-65-4320"
})

print(find_result)
# end-find-document

encrypted_client.close()

File: content/manual/versions/current/source/includes/qe-tutorials/node/queryable-encryption-tutorial.js
import "dotenv/config";
import { MongoClient } from "mongodb";
import * as qeHelper from "./queryable-encryption-helpers.js";

async function runExample() {
  // start-setup-application-variables
  const kmsProviderName = "<Your KMS Provider Name>";

  const uri = process.env.MONGODB_URI; // Your connection URI

  const keyVaultDatabaseName = "encryption";
  const keyVaultCollectionName = "__keyVault";
  const keyVaultNamespace = `${keyVaultDatabaseName}.${keyVaultCollectionName}`;
  const encryptedDatabaseName = "medicalRecords";
  const encryptedCollectionName = "patients";
  // end-setup-application-variables

  const kmsProviderCredentials =
    qeHelper.getKMSProviderCredentials(kmsProviderName);
  const customerMasterKeyCredentials =
    qeHelper.getCustomerMasterKeyCredentials(kmsProviderName);

  const autoEncryptionOptions = await qeHelper.getAutoEncryptionOptions(
    kmsProviderName,
    keyVaultNamespace,
    kmsProviderCredentials
  );

  // start-create-client
  const encryptedClient = new MongoClient(uri, {
    autoEncryption: autoEncryptionOptions,
  });
  // end-create-client

  await qeHelper.dropExistingCollection(encryptedClient, encryptedDatabaseName);
  await qeHelper.dropExistingCollection(encryptedClient, keyVaultDatabaseName);

  // start-encrypted-fields-map
  const encryptedFieldsMap = {
    encryptedFields: {
      fields: [
        {
          path: "patientRecord.ssn",
          bsonType: "string",
          queries: { queryType: "equality" },
        },
        {
          path: "patientRecord.billing",
          bsonType: "object",
        },
      ],
    },
  };
  // end-encrypted-fields-map

  const clientEncryption = qeHelper.getClientEncryption(
    encryptedClient,
    autoEncryptionOptions
  );

  await qeHelper.createEncryptedCollection(
    clientEncryption,
    encryptedClient.db(encryptedDatabaseName),
    encryptedCollectionName,
    kmsProviderName,
    encryptedFieldsMap,
    customerMasterKeyCredentials
  );

  // start-insert-document
  const patientDocument = {
    patientName: "Jon Doe",
    patientId: 12345678,
    patientRecord: {
      ssn: "987-65-4320",
      billing: {
        type: "Visa",
        number: "4111111111111111",
      },
      billAmount: 1500,
    },
  };

  const encryptedCollection = encryptedClient
    .db(encryptedDatabaseName)
    .collection(encryptedCollectionName);

  const result = await encryptedCollection.insertOne(patientDocument);
  // end-insert-document

  if (result.acknowledged) {
    console.log("Successfully inserted the patient document.");
  }

  // start-find-document
  const findResult = await encryptedCollection.findOne({
    "patientRecord.ssn": "987-65-4320",
  });
  console.log(findResult);
  // end-find-document

  await encryptedClient.close();
}

runExample().catch(console.dir);
